### ISSUE TITLE
[Bug] Adding ARCore pod breaks Custom Analytics Reporting on iOS device

### ISSUE THREAD

**Issue by adampsst** [firebase/quickstart-unity#1071]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2020.1.17f1
  * Firebase Unity SDK version: 7.2.0
  * Source you installed the SDK: UPM (.tgz)
  * Problematic Firebase Component: Analytics
  * Other Firebase Components in use: App, Crashlytics
  * Additional SDKs you are using: N/A
  * Platform you are using the Unity editor on: Mac
  * Platform you are targeting: iOS/Android
  * Scripting Runtime: IL2CPP (iOS), Mono (Android)

### [REQUIRED] Please describe the issue here:

When building to iOS, custom analytics are not being captured on the Firebase dashboard. However on Android they are. The only events on iOS which are captured are the default Firebase events, no custom evert data is showing. I have tried enablling Debug mode to check to see if they appear and it still does not show the data, again only the default Firebase events. We also have a test event being fired when the Firebase SDK's have been initialised and that is also not being captured on the dashboard (or debug view).

#### Steps to reproduce:

This occurs for all builds of our project on iOS, I have tried two different Firebase projects linked to Unity and neither show the data coming through.

#### Relevant Code:

Here is the initialisation code we are using:
```
using Cysharp.Threading.Tasks;
using Firebase.Analytics;
using Firebase.Crashlytics;
using Firebase.Extensions;
using UnityEngine;
using UnityEngine.SceneManagement;

namespace Mark.Utils
{
    public class FirebaseInit : MonoBehaviour
    {
        #region Serialized Fields

        [SerializeField] private string _sceneToLoad;

        #endregion

        #region Unity Methods

        private void Start()
        {
            Firebase.FirebaseApp.LogLevel = Firebase.LogLevel.Verbose;

            // Initialize Firebase
            Firebase.FirebaseApp.CheckAndFixDependenciesAsync().ContinueWithOnMainThread(task =>
            {
                var dependencyStatus = task.Result;
                if (dependencyStatus == Firebase.DependencyStatus.Available)
                {
                    var app = Firebase.FirebaseApp.DefaultInstance;
                    FirebaseAnalytics.SetAnalyticsCollectionEnabled(true);
                    Crashlytics.IsCrashlyticsCollectionEnabled = true;
                 
                    FirebaseAnalytics.LogEvent("TEST_EVENT", "test_param", "test_value");
                    
                    Debug.Log("Firebase SDK Setup");
                }
                else
                {
                    Debug.LogError($"Could not resolve all Firebase dependencies: {dependencyStatus}");
                }

                UniTask.Run(LoadScene);
            });
        }

        #endregion

        #region Private Methods

        /// <summary>
        /// Loads the next scene once we have initialized
        /// </summary>
        private async UniTask LoadScene()
        {
            await UniTask.SwitchToMainThread();
            Debug.Log($"[FirebaseInit] Loading scene: {_sceneToLoad}");
            SceneManager.LoadScene(_sceneToLoad);
        }

        #endregion
    }
}

```

**Comment by paulinon** [comment]
Hi @adampsst,

I noticed that your to log an event within the Start() function. It's possible that the event is not being reported because the FirebaseApp is not initiated yet. Naturally, the event cannot be enforced when you instantiate the object during gameplay.

With that, could you try the quickstart and see if the events aren't being reported from an iOS device?

**Comment by adampsst** [comment]
> Hi @adampsst,
> 
> I noticed that your to log an event within the Start() function. It's possible that the event is not being reported because the FirebaseApp is not initiated yet. Naturally, the event cannot be enforced when you instantiate the object during gameplay.
> 
> With that, could you try the quickstart and see if the events aren't being reported from an iOS device?

Yeah, I will set some time aside to try this in the quickstart project as well. The even you can see in the code above is actually just there as a test, we have around 15-20 more events that are not being captured either throughout the project, but are on Android.

**Comment by adampsst** [comment]
Hi,

I have tested the quickstart project and I can see events coming through on iOS (custom events), but still not in our own project (only default events that the SDK logs as mentioned before).

Any suggestions where to go from here? Would a list of other packages we are using help?

Adam

**Comment by adampsst** [comment]
It would appear the issue is related to having ARCore Extentions and the Firebase SDK's in the same project. Once they are both added in I am no longer able to fire custom events, they never get logged. I can also see that there are errors relating to classes being implemented in multiple frameworks, which only occurs when the ARCore packages are added in.

To also be clear, we are part of the partner program with Google using the ARCorePartner versions.

**Comment by adampsst** [comment]
> Since there haven't been any recent updates here, I am going to close this issue.
> 
> @adampsst if you're still experiencing this problem and want to continue the discussion just leave a comment here and we are happy to re-open this.

Please re-open

**Comment by paulinon** [comment]
Hi @adampsst,

Could you try using version 8.0.0 of the SDK and see if this resolves the issue?

**Comment by adampsst** [comment]
> Hi @adampsst,
> 
> Could you try using version 8.0.0 of the SDK and see if this resolves the issue?

I have tried this version and it is still not working. I have attached a file which contains various build errors inside xcode once i have Firebase products and ARCore Extensions installed in the same project.

[Xcode Build Errors.txt](https://github.com/firebase/quickstart-unity/files/6862854/Xcode.Build.Errors.txt)

**Comment by chkuang-g** [comment]
Ah interesting. Did you install both Firebase and ARCore using `.tgz`? Like the instruction described [here](https://developers.google.com/ar/develop/unity-arf/enable-arcore)?

And what is the version number of ARCorePartner version?

Here is my observation:
1. ARCore Extension [tgz](https://github.com/google-ar/arcore-unity-extensions/releases/) includes all the EDM4U libraries in its package. As a result, this blocks you from installing different version of EDM4U properly due to duplicated GUID. The latest ARCore includes EDM4U 1.2.162. To verify if this is the same to your project, check if `${YOUR_PROJECT_ROOT}/Library/PackageCache/com.google.ar.core.arfoundation.extensions@${SOME_RANDOM_STRING}/Editor/ExternalDependencyManager` exists.
2. [EDM4U 1.2.164](https://github.com/googlesamples/unity-jar-resolver/releases/tag/v1.2.164) added options to iOS Resolver to handle Swift Framework properly.  In short, the older version EDM4U will generate Podfile like
   ```
   target 'UnityFramework' do
     pod 'Firebase/Core', '7.11.0'
   end
   ```
   And the newer version (1.2.164+) will generate Podfile like the following:
   ```
   target 'UnityFramework' do
     pod 'Firebase/Core', '7.11.0'
   end
   target 'Unity-iPhone' do
   end
   use_frameworks!
   ```
   Just curious about which version of EDM4U is in your project.  Could you tell us the version number you saw on top of the settings windows using `Assets > External Dependency Manager > iOS Resolver > Settings` menu item?
   ![Screen Shot 2021-07-22 at 6 51 56 PM](https://user-images.githubusercontent.com/31869252/126729073-b75cac82-9a32-41e0-af93-2b0220b1ae4a.png)
3. Due to the point 1 and point 2, I think the only way to make this works is probably install using `.unitypackage`.  Like Firebase SDK from [here](https://developers.google.com/unity/archive#firebase) and ARCore ones from [here](https://developers.google.com/ar/develop/unity)

Could you verify if either the removal of ARCore extension package works or installing everything using `.unitypackage` works?

If so, this is unfortunately.  Since you are ARCore partner, I would recommend you to escalate this to ARCore team.  I will escalate using internal channel as well.

Please let us know!  Thank you.

**Comment by adampsst** [comment]
@chkuang-g Thanks for the long response. 

So to be clear I am having the same issue with both the partner SDK version of ARCore as well as the current public release version. However the installation for these is slightly different:

**Public Version**
This is installed by placing the pacakge into a GooglePackages folder, and then updating the manifest to point to the tgz file manually.

**Partner Version**
This is installed by placing the extracted SDK inside its own folder under the Packages folder inside the project.

Now, in either case, I am able to fix the EDM4U issue, and I am able to build without issue. However, as linked above, we get a lot of errors relating to duplicate classes on iOS, which I believe is the root cause of the issue.

For reference, we are using 1.2.166 EDM4U.

I have struck up a conversation with ARCore Partner team as well, if you can push this along that would be great. But the fact this is also causing issues with the public release version is something we need to resolve.

**Comment by chkuang-g** [comment]
Could you post the content in your `Podfile` and `Podfile.lock`?

Also, did you see any error if you run `pod install` under your Xcode project?

**Comment by adampsst** [comment]
POD file contents below. No errors whilst installing, `pod install` completes without issue as far as I can tell.

```
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '13.0'

target 'UnityFramework' do
  pod 'ARCorePartner/CloudAnchorsPartner', :path => './../../../ARCorePartner/'
  pod 'Firebase/Analytics', '8.1.1'
  pod 'Firebase/Core', '8.1.1'
  pod 'Firebase/Crashlytics', '8.1.1'
  pod 'Firebase/RemoteConfig', '8.1.1'
end
target 'Unity-iPhone' do
end
use_frameworks!
```

**Comment by adampsst** [comment]
Also, as mentioned earlier, this occurs when using the quickstart project along with the public ARCore Extensions SDK. Could someone in your team look at replicating this which may help resolve the issue?

**Comment by chkuang-g** [comment]
@adampsst 
Interesting.  So you are using later version EDM4U. I will started some investigation this week.

A couple of question about your project
- Which Xcode version are you using?
- What order did you use to install ARCore Extension SDK and Firebase? When such conflict happened, the order actually matter since the asset with the same GUID installed later will be ignored by Unity.
  I assume your order is the following.  But please confirm the order and the version when you reproduced this with the FCM quickstart.
  - `com.google.external-dependency-manager:1.2.167`
  - `com.google.firebase.app:8.0.0`
  - `com.google.firebase.analytics:8.0.0`
  - `com.google.firebase.crashlytics:8.0.0`
  - `com.google.ar.core.arfoundation.extensions:1.25.0`
- Lastly, could you include the `Podfile` when you attempted to reproduce the FCM quickstart with public version of ARCore? 

Thank you,
Shawn

**Comment by chkuang-g** [comment]
@adampsst 

**TL;DR;** I finally was able to reproduce the issue you described. I think the issue is with ARCore's pod, since by removing it, the Analytics event worked as intended. Also I found a workaround with a caveat: remove the following block in `Podfile` and run `pod install` again.
```
target 'Unity-iPhone' do
end
use_frameworks!
```
To prevent iOS resolver from adding this block back, disable the following settings in `Assets > External Dependency Manager > iOS Resolver > Settings` menu item.
![Screen Shot 2021-07-28 at 6 31 25 PM](https://user-images.githubusercontent.com/31869252/127417707-ae786c42-a1f5-4f81-bd19-f9e16c438b04.png)

**CAVEAT**: By doing so, your project will NOT be compatible to some other Unity packages which uses Swift framework, ex. Facebook Unity SDK. See #862

Unfortunately we have not crack the core reason for that yet.  That is why EDM4U added the option to add the following block to `Podfile` by default.
```
target 'Unity-iPhone' do
end
use_frameworks!
```
And we noticed that Admob and ARCore Pod spec is not compatible to these configuration at all. For some reason, that incompatibility broke Firebase Analytics custom event reporting.

I think ARCore team and Admob probably should update their Pod to make it works.  At this point, the workaround with caveat is probably the only option you have, or downgrade your Unity to 2019.2 or below before Unity introduce multi-target in Xcode project.

Let me contact our iOS team.

# Reproduction steps:
1. Use Firebase Unity SDK `8.1.0` and Unity 2020+.  It does not matter Firebase is installed using `.unitypackage` or `.tgz`. I was using Unity 2020.1.17f1 as @adampsst specified in the first post. But I think this should be reproducible using Unity 2019.3+.
1. Configure and build [Firebase Analytics testapp](https://github.com/firebase/quickstart-unity/tree/master/analytics/testapp) as described in readme.md
1. Enable DebugView as described [here](https://firebase.google.com/docs/analytics/get-started?platform=ios#start_logging_events). I noticed that I need to add both `-FIRAnalyticsDebugEnabled` and `-FIRDebugEnabled`
1. Open Debug View page on Firebase console.
1. Build Xcode project and launch the app.
1. Click on `Log Login` button.  Observe the event `login` show up on Firebase console.  It usually takes less than 1 minutes.

Then let's start to break it.
1. Uninstall the app from the iOS device.
1. Modify `Podfile` to add `ARCore/CloudAnchors` like the following.
  ```
  target 'UnityFramework' do
    pod 'Firebase/Analytics', '8.3.0'
    pod 'Firebase/Core', '8.3.0'
    pod "ARCore/CloudAnchors", '~> 1.25.0'
  end
  target 'Unity-iPhone' do
  end
  use_frameworks!
  ```
1. In a command line tool like iTerm, run `pod install`.
1. Build Xcode project and and launch the app.
1. Observe similar log attached in https://github.com/firebase/quickstart-unity/issues/1071#issuecomment-884951200
1. Click on `Log Login` button.  
1. Note that different installation is a different `Debug Device`. You should see the device count increased. Use the dropdown menu to switch to different device.
  ![Screen Shot 2021-07-28 at 7 08 44 PM](https://user-images.githubusercontent.com/31869252/127421087-c1296fef-d334-4a49-9c74-82156ec4cfd6.png)
1. Observe the event like `screen_view` is logged but NOT `login`.

Workaround
1. Uninstall the app from the iOS device.
2. Modify `Podfile` to remove `target 'Unity-iPhone'` and `use_frameworks!`, like the following:
  ```
  target 'UnityFramework' do
    pod 'Firebase/Analytics', '8.3.0'
    pod 'Firebase/Core', '8.3.0'
    pod "ARCore/CloudAnchors", '~> 1.25.0'
  end
  ```
1. In a command line tool like iTerm, run `pod install`.
1. Build Xcode project and and launch the app.
1. Click on `Log Login` button.  
1. Note that different installation is a different `Debug Device`. You should see the device count increased. Use the dropdown menu to switch to a different device.
1. Observe the event `login` is logged properly along with default Firebase events like `screen_view`.

**Comment by chkuang-g** [comment]
Hi @adampsst 

We did a full investigation. Thanks to @paulb777!

TL;DR; The workaround for you is simple: link pods statically. Analytics and ARCore pods are static binary libraries which does not work very well with the Xcode project generated by Unity after `2019.3`.  Static binary pods are pretty common and are not going anyway anytime soon.  Therefore, your best bet is to make sure pods are linked statically.

You can turn on the settings in `Assets > External Dependency Manager > iOS Resolver > Settings`
<img width="403" alt="Screen Shot 2021-08-09 at 7 47 21 PM" src="https://user-images.githubusercontent.com/31869252/128800745-2ad218a7-e4d5-4b16-bcda-fe177d5f21d1.png">

Once checked, you should see this line in the newly generated `Podfile`
```
use_frameworks! :linkage => :static
```

This means the pods should be linked statically.  This should work no matter pods are static libraries or dynamic libraries.

## Troubleshooting

1. We found the some other pods like `FBSDKCoreKit` from Facebook does not work very well with `use_frameworks! :linkage => :static` and you may see error like `Undefined Symbol`. What you need to do is to add a `Dummy.swift` to `UnityFramework` with the content like the following:
   ```
   import Foundation
   ```
   This will force Xcode project to link Swift libraries.
1. We found that Facebook Unity SDK would forcefully add `use_frameworks!` to `Podfile` no matter there is such line presented already or now.  See [code](https://github.com/facebook/facebook-sdk-for-unity/blob/4456caa9d40f31568297d7b5c92b354a9380da71/Facebook.Unity.Editor/FacebookPostprocess.cs#L33). As a result, you will see duplicated lines like the following:
   ```
   use_frameworks! :linkage => :static
   use_frameworks!
   ```
   Since latter line will take effect, this will make Xcode project links pods dynamically again. Your best coarse of action is to remove the second line manually.  I will file a bug to Facebook SDK to get this fixed.

That's it.  Let us know if you are still experiencing the issue with these workarounds.  I am thinking about adding some additional logic in EDM4U to turn on static linkage automatically.  But that will be a FR from EDM4U side.
